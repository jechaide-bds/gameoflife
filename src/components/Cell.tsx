import { Color } from "../utils/constants"

interface CellProps {
    grid: number[][]
    column: number
    row: number
}
const Cell = ({ grid, column, row } : CellProps) => {
    return           <div
    style={{
      width: 20,
      height: 20,
      backgroundColor: grid[column][row] ? Color.LIVING_CELL : undefined,
      border: Color.BORDER,
    }}
  />
}

export { Cell }