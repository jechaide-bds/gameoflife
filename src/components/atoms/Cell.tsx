import { setColorByState } from "../../utils/helpers"

interface CellProps {
    grid: number[][]
    column: number
    row: number
}
const Cell = ({ grid, column, row } : CellProps) => {
    return <div
            data-testid="cell-item"
            style={{
              width: 20,
              height: 20,
              backgroundColor: setColorByState(grid[column][row]),
              border: "1px solid #595959",
            }}
          />
}

export { Cell }