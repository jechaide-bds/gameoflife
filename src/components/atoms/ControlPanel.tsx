
export type GridFunction = () => React.SetStateAction<number[][]>

interface ControlPanelProps {
    setRunning: (running: boolean) => void
    running: boolean
    generateEmptyGrid: GridFunction
    randomTiles: GridFunction
    setGrid: (value: React.SetStateAction<number[][]>) => void
}
interface ButtonProps {
  onClick: () => void
  label: string
}

const Button = ({ onClick, label }: ButtonProps) => {
  return (
    <button onClick={onClick}
    style={{
      padding: '1.5em',
      margin: '2px',
      width: '8vw'
    }}>
      {label}
    </button>
  );
};

export default Button;

const ControlPanel = ({
  setRunning,
  running,
  generateEmptyGrid,
  randomTiles,
  setGrid
}: ControlPanelProps) => {
  return (
    <div
      style={{
        display: "flex",
        marginLeft: "20em",
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center",
      }}
    >
      <Button
        onClick={() => {
          setRunning(!running);
        }}
        label={running ? "Stop" : "Start"}
      />
      <Button
        onClick={() => {
          setGrid(generateEmptyGrid());
        }}
        label="Clear"
      />
      <Button
        onClick={() => {
          setGrid(randomTiles());
        }}
        label="Random"
      />
    </div>
  );
};

export { ControlPanel };
