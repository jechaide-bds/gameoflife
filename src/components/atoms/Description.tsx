const Title = () => {
    return <p
    style={{
        marginLeft: "23em",
        fontSize: "2em",
    }}> Conway Game of life </p>
}

const Rules = () => {
        return <div
                style={{
                textAlign: 'justify',
                justifyContent: 'center',
                width: 'fit-content',
                }}>
                    <p>Any live cell with fewer than two live neighbours dies, as if by underpopulation.
                        Any live cell with more than three live neighbours dies, as if by overpopulation.
                        Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
                    </p>
                </div>
}

const Description = () => {
   return <div 
    style={{
        textAlign: 'center',
        justifyContent: 'center',
        width: 'fit-content',
        }}>
        <Title />
    </div>
}
export { Title, Rules, Description }