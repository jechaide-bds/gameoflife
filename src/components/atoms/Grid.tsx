import { Cell } from "./Cell"

interface GridProps {
    grid: number[][] 
}
const Grid = ({ grid } : GridProps) => {

    return <div data-testid="grid-parent">
                {grid?.map((rows: number[], i) =>
                    rows.map((col, k) => (
                    <Cell data-testid="cell-item" key={i+k} grid={grid} column={i} row={k}/>
                    ))
                )}
            </div>
}

export { Grid }