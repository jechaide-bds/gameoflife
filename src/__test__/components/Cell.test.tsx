/* eslint-disable testing-library/prefer-screen-queries */
import { render } from "@testing-library/react";
import { Cell } from "../../components/atoms/Cell";
import { mockGridData } from "../mock/grid";

describe('Cell', () => {
    it('renders the grid container correctly', () => {
        const mock = {
            column: 1,
            row: 1
        }
        const { getAllByTestId } = render(
        <Cell 
            grid={mockGridData} 
            column={mock.column} 
            row={mock.row} 
        />
        );
        const cellItem = getAllByTestId("cell-item")

        expect(cellItem).toBeDefined();
    });
  });