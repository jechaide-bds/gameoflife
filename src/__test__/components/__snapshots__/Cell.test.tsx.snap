// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders the grid container correctly 1`] = `
Array [
  <div
    data-testid="cell-item"
    style="width: 20px; height: 20px; background-color: rgb(249, 89, 89); border: 1px solid #595959;"
  />,
]
`;
